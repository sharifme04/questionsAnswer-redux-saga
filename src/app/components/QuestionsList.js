import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import uuid from "uuid";

const QuestionsList = props => {
  const {
    questions,
    handleClick,
    selectedCorrect,
    selectedIncorrect,
    updateResults,
    totalQuestions
  } = props;

  return (
    <div className="mt-1">
      <h2>Deevio questions app</h2>
      <hr />
      {questions && questions.isLoading && questions.isLoading ? (
        <div className="text-center">
          <div className="spinner-border custom-spin text-info" role="status">
            <span className="sr-only">Loading...</span>
          </div>
        </div>
      ) : (
        <div className="row justify-content-center">
          {updateResults && updateResults ? (
            updateResults.map((e, i) => (
              <div
                key={uuid.v4()}
                className="col-xs-12 col-sm-12 col-md-6 mt-2"
              >
                <div className="card">
                  <div className="card-body">
                    <p className="card-title">
                      <b>{e.question}</b>
                    </p>
                    {e.allAnswers.sort().map(answer => (
                      <p
                        className={`card-text text-center custom-answer
                        ${
                          selectedCorrect &&
                          selectedCorrect.some(x => x[i] === answer)
                            ? "correct-color"
                            : selectedIncorrect &&
                              selectedIncorrect.some(x => x[i] === answer)
                            ? "incorrect-color"
                            : ""
                        }`}
                        onClick={() =>
                          handleClick(answer, i, e.question, e.correct_answer)
                        }
                        key={uuid.v4()}
                      >
                        {answer}
                      </p>
                    ))}
                  </div>
                </div>
              </div>
            ))
          ) : (
            <h5> Questions are not available</h5>
          )}
          <div className="mt-3 mb-4">
            <hr />
            {totalQuestions && (
              <Link to="/results">
                <button
                  type="button"
                  className="btn btn-success"
                  disabled={selectedCorrect.length !== totalQuestions}
                >
                  {selectedCorrect.length !== totalQuestions
                    ? "Please answer all question to go Next"
                    : "Submit Now!!"}
                </button>
              </Link>
            )}
          </div>
        </div>
      )}
    </div>
  );
};

QuestionsList.propTypes = {
  props: PropTypes.object,
  questions: PropTypes.object,
  selectedIncorrect: PropTypes.array,
  selectedCorrect: PropTypes.array,
  updateResults: PropTypes.array,
  handleClick: PropTypes.func
};

export default QuestionsList;
