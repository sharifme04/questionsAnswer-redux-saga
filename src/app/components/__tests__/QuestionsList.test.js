import React from "react";
import { shallow } from "enzyme";
import QuestionsList from "../QuestionsList";

jest.mock("uuid", () => {
  return {
    v4: jest.fn(() => 1)
  };
});

const fakeQuestions = {
  isLoading: false,
  questions: {
    response_code: 0,
    results: [
      {
        category: "Entertainment: Film",
        type: "multiple",
        difficulty: "easy",
        question:
          "Who played Deputy Marshal Samuel Gerard in the 1993 film &quot;The Fugitive&quot;?",
        correct_answer: "Tommy Lee Jones",
        incorrect_answers: ["Harrison Ford", "Harvey Keitel", "Martin Landau"]
      }
    ]
  }
};
const updateResults = [
  {
    category: "Entertainment: Film",
    type: "multiple",
    difficulty: "easy",
    question:
      "Who played Deputy Marshal Samuel Gerard in the 1993 film &quot;The Fugitive&quot;?",
    correct_answer: "Tommy Lee Jones",
    incorrect_answers: ["Harrison Ford", "Harvey Keitel", "Martin Landau"],
    allAnswers: [
      "Harrison Ford",
      "Harvey Keitel",
      "Martin Landau",
      "Tommy Lee Jones"
    ]
  }
];

describe("<QuestionsList />", () => {
  test("renders", () => {
    const component = shallow(<QuestionsList updateResults={updateResults} questions={fakeQuestions} handleClick={() => {}} />);
    expect(component).toMatchSnapshot();
  });

  test("return the default array when there is no data to map through", () => {
    const component = shallow(<QuestionsList />);
    expect(component).toMatchSnapshot();
  });

  test("does not break without value", () => {
    const component = shallow(<QuestionsList />);
    expect(component.find("h5")).toHaveLength(1);
    expect(component.find("p")).toHaveLength(0);
  });

  test("does not break with an empty array", () => {
    const component = shallow(<QuestionsList results={[]} />);
    expect(component.find("h5")).toHaveLength(1);
    expect(component.find("p")).toHaveLength(0);
  });
});
