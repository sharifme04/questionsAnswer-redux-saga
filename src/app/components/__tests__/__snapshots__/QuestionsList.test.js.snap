// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionsList /> renders 1`] = `
<div
  className="mt-1"
>
  <h2>
    Deevio questions app
  </h2>
  <hr />
  <div
    className="row justify-content-center"
  >
    <div
      className="col-xs-12 col-sm-12 col-md-6 mt-2"
      key="1"
    >
      <div
        className="card"
      >
        <div
          className="card-body"
        >
          <p
            className="card-title"
          >
            <b>
              Who played Deputy Marshal Samuel Gerard in the 1993 film &quot;The Fugitive&quot;?
            </b>
          </p>
          <p
            className="card-text text-center custom-answer
                        "
            key="1"
            onClick={[Function]}
          >
            Harrison Ford
          </p>
          <p
            className="card-text text-center custom-answer
                        "
            key="1"
            onClick={[Function]}
          >
            Harvey Keitel
          </p>
          <p
            className="card-text text-center custom-answer
                        "
            key="1"
            onClick={[Function]}
          >
            Martin Landau
          </p>
          <p
            className="card-text text-center custom-answer
                        "
            key="1"
            onClick={[Function]}
          >
            Tommy Lee Jones
          </p>
        </div>
      </div>
    </div>
    <div
      className="mt-3 mb-4"
    >
      <hr />
    </div>
  </div>
</div>
`;

exports[`<QuestionsList /> return the default array when there is no data to map through 1`] = `
<div
  className="mt-1"
>
  <h2>
    Deevio questions app
  </h2>
  <hr />
  <div
    className="row justify-content-center"
  >
    <h5>
       Questions are not available
    </h5>
    <div
      className="mt-3 mb-4"
    >
      <hr />
    </div>
  </div>
</div>
`;
